rclcpp/include/rclcpp/detail/rmw_implementation_specific_publisher_payload.hpp

```diff
@@ -0,0 +1,48 @@
+// Copyright 2019 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef RCLCPP__DETAIL__RMW_IMPLEMENTATION_SPECIFIC_PUBLISHER_PAYLOAD_HPP_
+#define RCLCPP__DETAIL__RMW_IMPLEMENTATION_SPECIFIC_PUBLISHER_PAYLOAD_HPP_
+
+#include "rcl/publisher.h"
+
+#include "rclcpp/detail/rmw_implementation_specific_payload.hpp"
+#include "rclcpp/visibility_control.hpp"
+
+namespace rclcpp
+{
+namespace detail
+{
+
+struct RCLCPP_PUBLIC RMWImplementationSpecificPublisherPayload
+  : public RMWImplementationSpecificPayload
+{
+  ~RMWImplementationSpecificPublisherPayload() override = default;
+
+  /// Opportunity for a derived class to inject information into the rcl options.
+  /**
+   * This is called after the rcl_publisher_options_t has been prepared by
+   * rclcpp, but before rcl_publisher_init() is called.
+   *
+   * By default the options are unmodified.
+   */
+  virtual
+  void
+  modify_rcl_publisher_options(rcl_publisher_options_t & rcl_publisher_options) const;
```

Comment:
Why?

Reply:
It is rmw implementation specific, that means it is not portable to just any rmw implementation, not that this class may only use the rmw API.

Reply:
Currently, the `modify_rcl_publisher_options` method may modify any field in `rcl_subscription_options_t` struct, for example `options.qos` or `options.allocator`.
I think it should be restricted to only modify `options.rmw_publisher_options`, and thus the argument could be directly a `rmw_publisher_options_t`.

I don't mind strongly if you address this comment or not, as this is not commonly used.

Reply:
Ok, I can narrow it, but I thought I might as well give this function as much access as is possible in case it needs it in the future. It maybe appropriate at some point to have it be able to change the QoS, for example if the QoS was being specified instead via a DDS QoS object in the payload.

Reply:
Changed in ac90f9a.